src/plugins/ui_core_views/row_size.ts

```diff
@@ -0,0 +1,164 @@
+  sizes: Immutable<Record<UID, Array<Pixel>>>;
+}
+
+export class UIRowSizePlugin extends UIPlugin<HeaderSizeState> implements HeaderSizeState {
+  static getters = ["getRowSize"] as const;
+
+  readonly sizes: Immutable<Record<UID, Array<Pixel>>> = {};
+
+  private ctx = document.createElement("canvas").getContext("2d")!;
+
+  handle(cmd: Command) {
+    switch (cmd.type) {
+      case "START":
+        for (const sheetId of this.getters.getSheetIds()) {
+          this.history.update("sizes", sheetId, this.computeRowSizes(sheetId));
+        }
+        break;
+      case "CREATE_SHEET": {
+        this.history.update("sizes", cmd.sheetId, this.computeRowSizes(cmd.sheetId));
+        break;
+      }
+      case "DUPLICATE_SHEET": {
+        const sizes = deepCopy(this.sizes[cmd.sheetId]);
+        this.history.update("sizes", cmd.sheetIdTo, sizes);
+        break;
+      }
+      case "DELETE_SHEET":
+        const sizes = { ...this.sizes };
+        delete sizes[cmd.sheetId];
+        this.history.update("sizes", sizes);
+        break;
+      case "REMOVE_COLUMNS_ROWS": {
+        if (cmd.dimension === "COL") return;
+        let sizes = [...this.sizes[cmd.sheetId]];
+        sizes = removeIndexesFromArray(sizes, cmd.elements);
+        this.history.update("sizes", cmd.sheetId, sizes);
+        break;
+      }
+      case "ADD_COLUMNS_ROWS": {
+        if (cmd.dimension === "COL") {
+          return;
+        }
+        const oldSizes = [...this.sizes[cmd.sheetId]];
+        const addIndex = getAddHeaderStartIndex(cmd.position, cmd.base);
+        const newRows = Array(cmd.quantity).fill(DEFAULT_CELL_HEIGHT);
+        const newSizes = insertItemsAtIndex(oldSizes, newRows, addIndex);
+        this.history.update("sizes", cmd.sheetId, newSizes);
+        break;
+      }
+      case "RESIZE_COLUMNS_ROWS":
+        {
+          const sheetId = cmd.sheetId;
+          if (cmd.dimension === "ROW") {
+            for (let el of cmd.elements) {
+              this.history.update("sizes", sheetId, el, this.getRowComputedSize(sheetId, el));
+            }
+          } else {
+            // Recompute row heights on col size change, they might have changed because of wrapped text
+            for (let row of range(0, this.getters.getNumberRows(sheetId))) {
+              this.history.update("sizes", sheetId, row, this.getRowComputedSize(sheetId, row));
+            }
+          }
+        }
+        break;
+      case "UPDATE_CELL":
+        const { sheetId, row } = cmd;
+        if (!this.getters.getUserRowSize(sheetId, row)) {
+          if (!this.sizes[sheetId]) debugger;
+          this.history.update("sizes", sheetId, row, this.getRowComputedSize(sheetId, row));
+        }
+        break;
+      case "ADD_MERGE":
+      case "REMOVE_MERGE":
+        const changerRows = new Set<number>();
+        for (let target of cmd.target) {
+          for (let row of range(target.top, target.bottom + 1)) {
+            changerRows.add(row);
+          }
```

Comment:
You can use a `JetSet` to avoid the extra for loop
```ts
const changerRows = new JetSet<number>();
for (const target of cmd.target) {
    changerRows.add(...range(target.top, target.bottom + 1))
}
```